src/multi/mod.rs

```diff
@@ -991,3 +998,194 @@ where
     Ok((input, res))
   }
 }
+
+/// Repeats the embedded parser and returns the results in a `Vec`.
+/// Fails if the amount of time the embedded parser is run is not
+/// within the specified range.
+/// # Arguments
+/// * `range` The amount of times to apply the parser.
+///   * A range with an upper bound `a..=b` limits the parser to run at most `b` times.
+///   * A range without an upper bound `a..` is equivalent to a range of `a..=usize::MAX`.
+///   * An empty range is invalid.
+/// * `parse` The parser to apply.
+/// ```rust
+/// # #[macro_use] extern crate nom;
+/// # use nom::{Err, error::ErrorKind, Needed, IResult};
+/// use nom::multi::many;
+/// use nom::bytes::complete::tag;
+///
+/// fn parser(s: &str) -> IResult<&str, Vec<&str>> {
+///   many(0..=2, tag("abc"))(s)
+/// }
+///
+/// assert_eq!(parser("abcabc"), Ok(("", vec!["abc", "abc"])));
+/// assert_eq!(parser("abc123"), Ok(("123", vec!["abc"])));
+/// assert_eq!(parser("123123"), Ok(("123123", vec![])));
+/// assert_eq!(parser(""), Ok(("", vec![])));
+/// assert_eq!(parser("abcabcabc"), Ok(("abc", vec!["abc", "abc"])));
+/// ```
+#[cfg(feature = "alloc")]
+#[cfg_attr(feature = "docsrs", doc(cfg(feature = "alloc")))]
+pub fn many<I, O, E, F, G, H>(
+  range: G,
+  mut parse: F,
+) -> impl FnMut(I) -> IResult<I, Vec<O>, E>
+where
+  I: Clone + InputLength,
+  F: Parser<I, O, E>,
+  E: ParseError<I>,
+  G: IntoRangeBounds<H>,
+  H: RangeBounds<usize>,
+{
+  let range = range.convert();
+  move |mut input: I| {
+    let start = match range.start_bound() {
+      Bound::Included(start) if !range.contains(start) => return Err(Err::Failure(E::from_error_kind(input, ErrorKind::Many))),
+      Bound::Excluded(start) if !range.contains(start) => return Err(Err::Failure(E::from_error_kind(input, ErrorKind::Many))),
+      Bound::Included(start) => Some(*start),
+      Bound::Excluded(start) => Some(*start + 1),
+      _ => None,
+    };
+    
+    let mut res = crate::lib::std::vec::Vec::with_capacity(start.unwrap_or(0));
+    
+    let mut parser = |count: usize| -> Option<Result<I, Err<E>>> {
+      let len = input.input_len();
+        match parse.parse(input.clone()) {
+          Ok((tail, value)) => {
+            // infinite loop check: the parser must always consume
+            if tail.input_len() == len {
+              return Some(Err(Err::Error(E::from_error_kind(input.clone(), ErrorKind::Many))));
+            }
+
+            res.push(value);
+            input = tail;
+            None
+          }
+          Err(Err::Error(e)) => {
+            if !range.contains(&count) {
+              return Some(Err(Err::Error(E::append(input.clone(), ErrorKind::Many, e))));
+            } else {
+              return Some(Ok(input.clone()));
+            }
+          }
+          Err(e) => {
+            return Some(Err(e));
+          }
+        }
+    };
+    
+    match range.end_bound() {
+      Bound::Unbounded => for count in (1..=core::usize::MAX).map(|x| x - 1) {
```

Comment:
don't make any sense, just use an infinite loop or a while let loop

Reply:
Assume the parser always succeeds. Having an infinite loop would basically break down to this:
```rust
loop {
  vec.push(result)
}
```
This would mean we could try to push more than `usize::MAX` elements onto the Vec. That would cause a panic and is therefore invalid. To prevent this an unbounded range is instead interpreted as having an upper bound of at most `usize::MAX` iterations.

Reply:
first this would panic at `isize::MAX`, secondly the program will ABORT before so what ? and third this is the impossible case cause Rust std choice to not return an result for allocation, use try_reserve api if you want avoid that, in any case, don't use this useless loop.

Reply:
Outdated. New implementation avoids having the parser map the iterators.

An infinte loop with a `vec.push` is incorrect so a fixed upper bound of `usize::MAX` is necessary.